# Machine Learning Spring 2025

> **Project 1 - Temperature Prediction**<br><br>Team: ST_ML2025_2

In [28]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re

In [4]:
# Load the given datasets(located in the input folder)
train_df = pd.read_csv("./input/train_dataset.csv")
test_df = pd.read_csv("./input/test_dataset.csv")
station_info_df = pd.read_csv("./input/station_info.csv")
sample_submission_df = pd.read_csv("./input/submission_sample.csv")

# Dataset 분석

주어진 데이터셋의 구성은 다음과 같다.

- `train_dataset.csv`: 동두천, 서울, 강화, 인천, 이천, 양평 관측소의 2019-2024년 데이터셋
- `test_dataset.csv`: 파주, 수원 관측소 데이터셋
- `station_info.csv`: 기상 관측소별 정보

In [14]:
train_df.sample(10)

id  station station_name   date  cloud_cover_0  cloud_cover_1  \
2215    4403      108           서울  01-29            0.0            0.0   
913      913       98          동두천  07-02            0.0            0.0   
4461    6649      112           인천  03-25            0.0            0.0   
8754   13133      201           강화  12-29            9.0            9.0   
3328    5516      108           서울  02-15            8.0            9.0   
13113  17492      203           이천  12-12            0.0            0.0   
9498   13877      202           양평  01-14            0.0            0.0   
3234    5422      108           서울  11-13            0.0            0.0   
10105  14484      202           양평  09-13            9.0            7.0   
115      115       98          동두천  04-26            9.0            0.0   

       cloud_cover_10  cloud_cover_11  cloud_cover_12  cloud_cover_13  ...  \
2215              0.0             0.0             0.0             0.0  ...   
913               3.0             3.0             2.0             0.0  ...   
4461              0.0             0.0             0.0             0.0  ...   
8754              0.0             0.0             0.0             1.0  ...   
3328              7.0             7.0             8.0             8.0  ...   
13113             0.0             0.0             0.0             0.0  ...   
9498              0.0             0.0             6.0             6.0  ...   
3234              1.0             0.0             7.0             3.0  ...   
10105             0.0             1.0             0.0             1.0  ...   
115               9.0            10.0             9.0             9.0  ...   

       wind_speed_23  wind_speed_3  wind_speed_4  wind_speed_5  wind_speed_6  \
2215             1.4           1.8           1.4           1.2           0.8   
913              0.7           0.2           0.3           0.1           0.9   
4461             4.0           4.8           5.0           3.8           3.8   
8754             1.4           0.5           0.8           0.5           0.8   
3328             3.5           4.0           4.1           3.9           3.5   
13113            0.8           0.1           0.6           0.3           0.1   
9498             0.2           1.5           0.4           0.5           1.4   
3234             1.0           1.0           0.6           2.1           2.1   
10105            1.4           0.5           0.1           0.3           0.0   
115              1.6           1.6           2.3           1.6           1.3   

       wind_speed_7  wind_speed_8  wind_speed_9  climatology_temp    target  
2215            1.5           1.0           1.1         -1.251786  3.051786  
913             0.9           0.5           0.8         24.908929 -3.408929  
4461            4.5           4.3           4.4         10.094643 -0.694643  
8754            0.7           0.3           0.6         -2.814286  6.814286  
3328            3.7           3.9           3.1         -2.257143 -5.042857  
13113           0.0           0.2           0.6         -0.032143  0.632143  
9498            0.7           0.7           0.4         -0.823214  2.923214  
3234            2.0           1.4           1.8          7.971429  3.328571  
10105           0.1           0.5           0.9         22.944643  1.155357  
115             2.0           1.3           0.6         13.637500 -2.637500  

[10 rows x 342 columns]

In [12]:
train_df.describe()

id       station  cloud_cover_0  cloud_cover_1  \
count  13132.000000  13132.000000   12945.000000   12920.000000   
mean    9484.110493    153.980658       2.915798       3.022291   
std     5311.954253     48.183220       3.646779       3.652165   
min        0.000000     98.000000       0.000000       0.000000   
25%     5470.750000    108.000000       0.000000       0.000000   
50%     8753.500000    112.000000       0.000000       1.000000   
75%    14227.250000    202.000000       6.000000       6.000000   
max    17510.000000    203.000000      10.000000      10.000000   

       cloud_cover_10  cloud_cover_11  cloud_cover_12  cloud_cover_13  \
count    12916.000000    12926.000000    12931.000000    12926.000000   
mean         3.126742        3.092063        3.046400        3.035046   
std          3.659422        3.589739        3.512063        3.483002   
min          0.000000        0.000000        0.000000        0.000000   
25%          0.000000        0.000000        0.000000        0.000000   
50%          1.000000        1.000000        1.000000        1.000000   
75%          7.000000        7.000000        6.000000        6.000000   
max         10.000000       10.000000       10.000000       10.000000   

       cloud_cover_14  cloud_cover_15  ...  wind_speed_23  wind_speed_3  \
count    12919.000000    12938.000000  ...   13122.000000  13120.000000   
mean         2.992182        2.911115  ...      -0.032754     -2.491814   
std          3.462070        3.440298  ...     123.464434    195.197628   
min          0.000000        0.000000  ...   -9999.000000  -9999.000000   
25%          0.000000        0.000000  ...       0.600000      0.400000   
50%          1.000000        1.000000  ...       1.200000      1.000000   
75%          6.000000        6.000000  ...       2.100000      1.900000   
max         10.000000       10.000000  ...      10.700000     11.300000   

       wind_speed_4  wind_speed_5  wind_speed_6  wind_speed_7  wind_speed_8  \
count  13116.000000  13112.000000  13119.000000  13120.000000  13125.000000   
mean      -0.234637     -1.011882     -4.068389     -7.119787    -10.073272   
std      123.489868    151.258218    230.949420    289.453706    337.895538   
min    -9999.000000  -9999.000000  -9999.000000  -9999.000000  -9999.000000   
25%        0.400000      0.400000      0.400000      0.400000      0.500000   
50%        0.900000      0.900000      0.900000      0.900000      1.000000   
75%        1.900000      1.800000      1.800000      1.800000      2.000000   
max        9.500000     10.600000     12.400000     11.100000     11.300000   

       wind_speed_9  climatology_temp        target  
count  13127.000000      13132.000000  13132.000000  
mean      -9.119029         12.658557      0.221979  
std      326.432094         10.023504      2.960544  
min    -9999.000000         -4.487500    -12.864286  
25%        0.700000          3.292857     -1.643052  
50%        1.300000         12.842857      0.157143  
75%        2.200000         22.271429      2.045536  
max        9.400000         28.455357     11.778571  

[8 rows x 340 columns]

In [13]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13132 entries, 0 to 13131
Columns: 342 entries, id to target
dtypes: float64(338), int64(2), object(2)
memory usage: 34.3+ MB


## Feature 분석

### Feature 분류 및 설명

In [40]:
def search_time_based_feature_names(df):
    "각 시간별로 구성된 feature들(e.g. 'cloud_cover_0', 'cloud_cover_1' ...)에 대해 각각의 고유한 이름(e.g. 'cloud_cover')을 찾는다."

    time_based_features_pattern = re.compile(r"^(.*)_(\d{1,2})$")
    searched_time_based_feature_names = set()

    for column in df.columns:
        match = time_based_features_pattern.match(column)
        if match:
            feature_name, hour = match.groups()
            hour = int(hour)
            searched_time_based_feature_names.add(feature_name)

    return searched_time_based_feature_names

time_based_feature_names = search_time_based_feature_names(train_df)

print(len(time_based_feature_names))
print(time_based_feature_names)

14
{'humidity', 'vapor_pressure', 'wind_direction', 'min_cloud_height', 'sunshine_duration', 'dew_point', 'local_pressure', 'sea_level_pressure', 'surface_temp', 'cloud_cover', 'snow_depth', 'visibility', 'wind_speed', 'precipitation'}


train dataset에 있는 feature를 특성에 따라 다음과 같은 카테고리별로 정리하였다.

1. Data 식별
   - `id`: 순서(identical)
   - `station`: 지상관측소 번호(98: "동두천", 201: "강화" 등)
   - `station_name`: 지상관측소 이름("동두천", "서울", "강화" 등)
   - `date`: 날짜(월-일 형식, 1월 29일은 "01-29"로 표기)

2. 구름 관련 feature
    - `cloud_cover_[0-23]`: 증하층운량(10분위, 0~10)
    - `min_cloud_height_[0-23]`: 시간별 최저운고(100m 단위)

3. 온도 관련 feature
    - `dew_point_[0-23]`: 시간별 이슬점 온도(°C)
    - `surface_temp_[0-23]`: 시간별 지면온도(°C)
    - `climatology_temp`: 해당 날짜의 평균 기온(°C) (7년 동안 평균)

4. 습도 관련 feature
    - `humidity_[0-23]`: 시간별 습도(%)
    - `vapor_pressure_[0-23]`: 시간별 증기압(hPa)
    - `percipitation_[0-23]`: 시간별 강수량(mm)
    - `snow_depth_[0-23]`: 시간별 적설(cm)

5. 기압 관련 feature
    - `local_pressure_[0-23]`: 시간별 현지기압(hPa)
    - `sea_level_pressure_[0-23]`: 시간별 해면기압(hPa)

6. 가시성 관련 feature
    - `visibility_[0-23]`: 시간별 시정(10m 단위)
    - `sunshine_duration_[0-23]`: 시간별 일조(hr)

7. 바람 관련 feature
    - `wind_speed_[0-23]`: 시간별 풍속(m/s)
    - `wind_direction_[0-23]`: 시간별 풍향(도 단위)

8. target feature
    - `target`: 시간별 기온(°C, 다음날 평균 기온(°C)에서 `climatology_temp`를 뺀 값)


### 결측치 패턴

1. `-9999`: 관측소 기계에서 감지한 결측치 또는 이상치
2. `NaN`:
    - `sunshine_duration`: 야간 시간대(0, 1, 2, 3, 4, 5, 22, 23시), 즉 해가 없는 시간대
    - `snow_depth`: 눈이 오지 않은 경우
    - `precipitation`: 비가 오지 않은 경우




## Feature Engineering

### 결측치 처리